augur/curate/format_dates.py

```diff
@@ -0,0 +1,237 @@
+
+
+def format_date(date_string, expected_formats):
+    """
+    Format *date_string* to ISO 8601 date (YYYY-MM-DD) by trying to parse it
+    as one of the provided *expected_formats*.
+
+    Parameters
+    ----------
+    date_string: str
+        Date string to format
+    expected_formats: list[str]
+        List of expected formats for the provided date string
+
+    Returns
+    -------
+    str or None:
+        Formatted date string or None if the parsing of the date string failed.
+        If *date_string* is an incomplete date, the date is masked with 'XX'.
+        Dates without year will be formatted as 'XXXX-XX-XX', even if month/day are known.
+        Dates without month will be formatted as 'YYYY-XX-XX', even if day is known.
+        Dates without day will be formatted as 'YYYY-MM-XX'.
+
+
+    >>> expected_formats = ['%Y', '%Y-%m', '%Y-%m-%d', '%Y-%m-%dT%H:%M:%SZ', '%m-%d']
+    >>> format_date("01-01", expected_formats)
+    'XXXX-XX-XX'
+    >>> format_date("2020", expected_formats)
+    '2020-XX-XX'
+    >>> format_date("2020-01", expected_formats)
+    '2020-01-XX'
+    >>> format_date("2020-1-15", expected_formats)
+    '2020-01-15'
+    >>> format_date("2020-1-1", expected_formats)
+    '2020-01-01'
+    >>> format_date("2020-01-15", expected_formats)
+    '2020-01-15'
+    >>> format_date("2020-01-15T00:00:00Z", expected_formats)
+    '2020-01-15'
+    """
+    # Set of directives that can be converted to complete date with year, month, and day
+    year_month_day_directives = {
+        # Locale's full date representation
+        ('%c',),('%x',),
+        # Dates with ISO 8601 week dates for year ('%G' is NOT interchangeable with '%Y'), ISO 8601 week ('%V'), and weekdays
+        ('%G', '%V', '%A'),('%G', '%V', '%a'),('%G', '%V', '%w'),('%G', '%V', '%u'),
+        # Dates with year, week, and weekday
+        ('%y', '%U', '%A'), ('%y', '%U', '%a'), ('%y', '%U', '%w'), ('%y', '%U', '%u'),
+        ('%y', '%W', '%A'), ('%y', '%W', '%a'), ('%y', '%W', '%w'), ('%y', '%W', '%u'),
+        ('%Y', '%U', '%A'), ('%Y', '%U', '%a'), ('%Y', '%U', '%w'), ('%Y', '%U', '%u'),
+        ('%Y', '%W', '%A'), ('%Y', '%W', '%a'), ('%Y', '%W', '%w'), ('%Y', '%W', '%u'),
+        # Dates with year and day of the year
+        ('%y', '%j'), ('%Y', '%j'),
+        # Dates with year, month, and day
+        ('%y', '%b', '%d'), ('%y', '%B', '%d'), ('%y', '%m', '%d'),
+        ('%Y', '%b', '%d'), ('%Y', '%B', '%d'), ('%Y', '%m', '%d'),
+    }
+
+    # Set of directives that can be converted to incomplete dates, missing the day
+    year_month_directives = {
+        ('%y', '%b'), ('%y', '%B'), ('%y', '%m'),
+        ('%Y', '%b'), ('%Y', '%B'), ('%Y', '%m'),
+    }
+
+    # Set of directives that can be converted to incomplete dates, missing the month and day
+    year_directives = {
+        ('%y',), ('%Y',)
+    }
+
+    for date_format in expected_formats:
+        try:
+            parsed_date = datetime.strptime(date_string, date_format)
```

Comment:
I was going to suggest `datetime.date.strftime` instead of `datetime.datetime.strptime`, but I noticed the `%c` directive, which includes time parts, is included. Is this intentional? The time part is left unused in the rest of the function here.

Reply:
Leaving the time parts of the `%c` directive unused is intentional since we don't expect times in downstream Augur commands.

Reply:
Oh, I meant that I wasn't sure if `%c` should be supported. But I understand how it can be useful – since year/month/day information still can be extracted from it, doesn't hurt to support it.